core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -160,7 +160,16 @@ public void putIntoWriteSet(Key key, Put put) {
   }
 
   public void putIntoDeleteSet(Key key, Delete delete) {
-    writeSet.remove(key);
+    Put put = writeSet.get(key);
+    if (put != null) {
+      if (put.isInsertModeEnabled()) {
+        throw new IllegalArgumentException(
+            CoreError.CONSENSUS_COMMIT_DELETING_ALREADY_INSERTED_DATA_NOT_ALLOWED.buildMessage());
+      }
+
+      writeSet.remove(key);
+    }
```

Comment:
Added validation to ensure that Delete is not executed on a record that has already been inserted in the same transaction.